# ClinVarome annotation

<img src="img/clinvarome_va_ga.png"  width="250" height="250">

## Overview

ClinVarome is a list of genes involved in human disease, based on ClinVar database extraction via `Genome Alert!`.
ClinVarome annotation allows enhanced ClinVarome with useful annotations for genetic diagnostics.

It provides for each gene:

- **Clinical features** extracted from clinVCF VCF file
- **Molecular consequences reported**
- **Date of first and last apparition of pathogenic variant**
- **Number of reported (likely) pathogenic variants**

Based on the following features, the annotation provides also a **clinical validity score** for genes generated by a Agglomerative Clustering model :

- Date between first and last apparition of pathogenic variant
- Number of reported (likely) pathogenic variants
- Highest review confidence variant status in (likely) pathogenic variant
- Highest ACMG classification variant status

ClinVarome annotation is developped in `python > 3.6`, and need `Genome Alert!` processed data.

## Quick start

You need to have installed Variant Alert! [https://github.com/SeqOne/variant_alert](https://github.com/SeqOne/variant_alert) and output VCFs from clinVCF in a folder [https://github.com/SeqOne/clinvcf](https://github.com/SeqOne/clinvcf).

```bash
# Go in the Variant Alert! repository and launch virtual environment
cd variant_alert/
poetry shell

# Process Variant Alert! data for every months
cd clinvarome/
python variant_alert_execution.py --vcf-path VCF_PATH --VA-output-path OUTPUT_PATH

# Annotate ClinVarome

python clinvarome_annotation.py --vcf VCF_FILE --clinvarome CLINVAROME_FILE_TO_ANNOTATE --compare-gene OUTPUT_PATH/compare-gene_total.tsv --compare-variant OUTPUT_PATH/compare-variant_total.tsv --output-dir OUTPUT_DIR/
```

## Usage

```bash
usage: clinvarome_annotation.py [-h] --vcf VCF_FILE --clinvarome CLINVAROME
                                --compare-gene COMPARE_GENE --compare-variant
                                COMPARE_VARIANT [--max-date MAX_DATE]
                                [--gnomad] [--manual-score] --output-dir
                                OUTPUT_DIR

Add additional annotations to the ClinVarome.

optional arguments:
  -h, --help            show this help message and exit
  --vcf VCF_FILE        vcf input from clinVCF
  --clinvarome CLINVAROME
                        ClinVarome file from Variant Alert! to be annotated.
  --compare-gene COMPARE_GENE
                        Variant Alert! compare gene concatenation file of all
                        previous ClinVar release.
  --compare-variant COMPARE_VARIANT
                        Variant Alert! compare variant concatenation file of
                        all previous ClinVar release.
  --gnomad              Add max pathogenic gnomAD filtered allele frequency
                        for each gene, if clinVCF is annotated with gnomAD v3
                        by vcfanno.
  --output-dir OUTPUT_DIR
                        Path to the output folder.
```

### Output format

ClinVarome annotation generate an enhanced ClinVarome TSV (file with additional suffix `_clinvarome_annotation.tsv`).

## Validation experiments

Two experiment notebooks are available that follow the upcoming publication of `Genome Alert!` (`Figure_1.ipynb` and `Figure_2.ipynb`).

---

_Genome Alert! is a collaboration of :_

[![SeqOne](img/logo-seqone.png)](https://seq.one/)

[![Universit√© Grenoble Alpes](img/logo-uga.png)](https://iab.univ-grenoble-alpes.fr/)

[![CHU de Rouen](img/logo-CHU.png)](https://www.chu-rouen.fr/service/service-de-genetique/)
