# ClinVarome annotation

<img src="img/clinvarome_va_ga.png"  width="250" height="250">

## Overview

ClinVarome is a list of genes involved in human disease, based on ClinVar database extraction via `Genome Alert!`.
ClinVarome annotation allows enhanced ClinVarome with useful annotations for genetic diagnostics.

It provides for each gene:

- **Clinical features** extracted from clinVCF VCF file
- **Molecular consequences reported**
- **Date of first and last apparition of pathogenic variant**
- **Number of reported (likely) pathogenic variants**

Based on the following features, the annotation provides also a **clinical validity score** for genes generated by a Agglomerative Clustering model :

- Date between first and last apparition of pathogenic variant
- Number of reported (likely) pathogenic variants
- Highest review confidence variant status in (likely) pathogenic variant
- Highest ACMG classification variant status

ClinVarome annotation is developped in `python > 3.6`, and need `Genome Alert!` processed data.

## Quick start

ClinVarome annotation need for a ClinVar release :

- a VCF file from [ClinVCF](https://github.com/SeqOne/clinvcf)
- a ClinVarome file from [Variant Alert!](https://github.com/SeqOne/variant_alert)
- a concatenated file of all [Variant Alert!](https://github.com/SeqOne/variant_alert) compare-variant output (compare-variant_total.tsv.gz)
- a concatenated file of all [Variant Alert!](https://github.com/SeqOne/variant_alert) compare-gene output (compare-gene_total.tsv.gz)

Run ClinVarome

```bash
poetry install
#For testing
mkdir OUTPUT_DIR
poetry run clinvarome/clinvarome_annotation.py \
  --vcf tests/full_data/clinvar_GRCh38_{latest}.vcf.gz \
  --clinvarome tests/full_data/clinvarome_{latest}.tsv \
  --compare-gene tests/full_data/compare-gene_total.tsv.gz \
  --compare-variant tests/full_data/compare-variant_total.tsv.gz \
  --output-dir OUTPUT_DIR/
```

## Usage

```bash
usage: clinvarome_annotation.py [-h] --vcf VCF_FILE --clinvarome CLINVAROME
                                --compare-gene COMPARE_GENE --compare-variant
                                COMPARE_VARIANT [--max-date MAX_DATE]
                                [--gnomad] [--manual-score] --output-dir
                                OUTPUT_DIR

Add additional annotations to the ClinVarome.

optional arguments:
  -h, --help              show this help message and exit
  --vcf        VCF_FILE   vcf input from clinVCF
  --clinvarome CLINVAROME
                          ClinVarome file from Variant Alert! to be annotated.
  --compare-gene COMPARE_GENE
                          Variant Alert! compare gene concatenation file of all
                          previous ClinVar release.
  --compare-variant COMPARE_VARIANT
                          Variant Alert! compare variant concatenation file of
                          all previous ClinVar release.
  --gnomad                Add max pathogenic gnomAD filtered allele frequency
                          for each gene, if clinVCF is annotated with gnomAD v3
                          by vcfanno.
  --output-dir OUTPUT_DIR
                          Path to the output folder.
```

### Output format

ClinVarome annotation generate an enhanced ClinVarome TSV (file with additional suffix `_clinvarome_annotation.tsv`).

---

_Genome Alert! is a collaboration of :_

[![SeqOne](img/logo-seqone.png)](https://seq.one/)

[![Universit√© Grenoble Alpes](img/logo-uga.png)](https://iab.univ-grenoble-alpes.fr/)

[![CHU de Rouen](img/logo-CHU.png)](https://www.chu-rouen.fr/service/service-de-genetique/)
